File: packages/agent-sdk/src/transport.ts
import { createRequire } from "node:module";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { NomaSpawnError, NomaTimeoutError, NomaTransportError } from "./errors.js";

const require_ = createRequire(import.meta.url);

export type StdioMcpClientOptions = {
  mcpServerBin?: string;
  env?: NodeJS.ProcessEnv;
  requestTimeoutMs?: number;
};

export type ToolDescriptor = { name: string };

export type ToolCallResult = {
  isError: boolean;
  text: string;
};

const DEFAULT_TIMEOUT_MS = 30_000;

function resolveServerBin(override?: string): string {
  if (override) return override;
  try {
    return require_.resolve("@ferax564/noma-mcp-server/dist/index.js");
  } catch (cause) {
    throw new NomaSpawnError("could not resolve @ferax564/noma-mcp-server binary", cause);
  }
}

export class StdioMcpClient {
  private readonly client: Client;
  private readonly transport: StdioClientTransport;
  private readonly timeoutMs: number;
  private closed = false;

  private constructor(client: Client, transport: StdioClientTransport, timeoutMs: number) {
    this.client = client;
    this.transport = transport;
    this.timeoutMs = timeoutMs;
  }

  static async spawn(options: StdioMcpClientOptions = {}): Promise<StdioMcpClient> {
    const bin = resolveServerBin(options.mcpServerBin);
    // StdioClientTransport expects env as Record<string, string> — NodeJS.ProcessEnv
    // is Record<string, string | undefined>, so we must drop undefined values
    // before passing through (or the strict-TS build fails).
    const env = options.env
      ? Object.fromEntries(
          Object.entries(options.env).filter(
            (entry): entry is [string, string] => typeof entry[1] === "string",
          ),
        )
      : undefined;
    const transport = new StdioClientTransport({
      command: process.execPath,
      args: [bin],
      ...(env ? { env } : {}),
    });
    const client = new Client(
      { name: "@ferax564/noma-agent-sdk", version: "0.1.0" },
      { capabilities: {} },
    );
    try {
      await client.connect(transport);
    } catch (cause) {
      throw new NomaSpawnError(`failed to start mcp-server: ${(cause as Error).message}`, cause);
    }
    return new StdioMcpClient(client, transport, options.requestTimeoutMs ?? DEFAULT_TIMEOUT_MS);
  }

  async listTools(): Promise<ToolDescriptor[]> {
    this.assertOpen();
    const res = await this.withTimeout(this.client.listTools());
    return res.tools.map((t) => ({ name: t.name }));
  }

  async callTool(name: string, args: Record<string, unknown>): Promise<ToolCallResult> {
    this.assertOpen();
    // The MCP SDK's callTool return type is a union: it can be the
    // CallToolResult shape (`{ content, isError? }`) OR a legacy
    // `{ toolResult }` envelope. Strict TS rejects direct field access.
    // The @ferax564/noma-mcp-server always emits the CallToolResult shape, but the
    // SDK's static type doesn't know that — narrow by sniffing for
    // `content`. Throw a transport error if the envelope is unexpected.
    const raw = await this.withTimeout(
      this.client.callTool({ name, arguments: args }),
    );
    if (!("content" in raw)) {
      throw new NomaTransportError(`tool ${name} returned legacy toolResult envelope; expected CallToolResult`);
    }
    const content = raw.content;
    if (!Array.isArray(content) || content.length === 0 || content[0]?.type !== "text") {
      throw new NomaTransportError(`tool ${name} returned no text content`);
    }
    return { isError: raw.isError === true, text: String(content[0].text) };
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    await this.client.close();
  }

  private assertOpen(): void {
    if (this.closed) throw new NomaTransportError("client closed");
  }

  private withTimeout<T>(p: Promise<T>): Promise<T> {
    return new Promise<T>((resolve, reject) => {
      const handle = setTimeout(
        () => reject(new NomaTimeoutError(`request exceeded ${this.timeoutMs}ms`)),
        this.timeoutMs,
      );
      p.then((v) => {
        clearTimeout(handle);
        resolve(v);
      }).catch((e) => {
        clearTimeout(handle);
        reject(e);
      });
    });
  }
}

File: packages/agent-sdk-py/src/noma_agent_sdk/transport.py
"""Stdio MCP transport — spawns the Node ``@ferax564/noma-mcp-server``.

Mirrors ``packages/agent-sdk/src/transport.ts``. Default command is
``npx -y @ferax564/noma-mcp-server`` so the Python SDK does not require the
caller to clone the Noma repo — the npm registry resolves the server binary.
Callers may override with a local node binary for development / CI.
"""

from __future__ import annotations

import asyncio
from contextlib import AsyncExitStack
from dataclasses import dataclass, field
from typing import Any

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

from .errors import NomaSpawnError, NomaTimeoutError, NomaTransportError

DEFAULT_TIMEOUT_S = 30.0

DEFAULT_COMMAND = "npx"
DEFAULT_ARGS: tuple[str, ...] = ("-y", "@ferax564/noma-mcp-server")


@dataclass
class StdioMcpClientOptions:
    """Configuration for spawning the Node MCP server child process."""

    command: str = DEFAULT_COMMAND
    args: list[str] = field(default_factory=lambda: list(DEFAULT_ARGS))
    env: dict[str, str] | None = None
    request_timeout_s: float = DEFAULT_TIMEOUT_S


@dataclass
class ToolCallResult:
    is_error: bool
    text: str


class StdioMcpClient:
    """Thin async wrapper around ``mcp.ClientSession`` over stdio.

    Use :meth:`spawn` (an ``async`` factory) to create — the session must be
    set up inside an event loop, and the constructor cannot be ``async``.

    Always call :meth:`close` (or use ``async with``) to tear down the child
    process. Leaking the process leaks file descriptors on every spawn.
    """

    def __init__(
        self,
        session: ClientSession,
        exit_stack: AsyncExitStack,
        timeout_s: float,
    ) -> None:
        self._session = session
        self._exit_stack = exit_stack
        self._timeout_s = timeout_s
        self._closed = False

    @classmethod
    async def spawn(cls, options: StdioMcpClientOptions | None = None) -> "StdioMcpClient":
        opts = options or StdioMcpClientOptions()
        params = StdioServerParameters(
            command=opts.command,
            args=list(opts.args),
            env=opts.env,
        )
        stack = AsyncExitStack()
        try:
            read, write = await stack.enter_async_context(stdio_client(params))
            session = await stack.enter_async_context(ClientSession(read, write))
            await session.initialize()
        except Exception as cause:
            await stack.aclose()
            raise NomaSpawnError(
                f"failed to start mcp-server ({opts.command} {' '.join(opts.args)}): {cause}",
                cause,
            ) from cause
        return cls(session, stack, opts.request_timeout_s)

    async def __aenter__(self) -> "StdioMcpClient":
        return self

    async def __aexit__(self, *_exc: object) -> None:
        await self.close()

    async def list_tools(self) -> list[str]:
        self._assert_open()
        res = await self._with_timeout(self._session.list_tools())
        return [t.name for t in res.tools]

    async def call_tool(self, name: str, arguments: dict[str, Any]) -> ToolCallResult:
        self._assert_open()
        raw = await self._with_timeout(
            self._session.call_tool(name, arguments=arguments)
        )
        content = getattr(raw, "content", None)
        if not content:
            raise NomaTransportError(f"tool {name} returned no content")
        first = content[0]
        text = getattr(first, "text", None)
        if text is None:
            raise NomaTransportError(f"tool {name} returned non-text content")
        return ToolCallResult(is_error=bool(getattr(raw, "isError", False)), text=str(text))

    async def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        await self._exit_stack.aclose()

    def _assert_open(self) -> None:
        if self._closed:
            raise NomaTransportError("client closed")

    async def _with_timeout(self, coro):
        try:
            return await asyncio.wait_for(coro, timeout=self._timeout_s)
        except asyncio.TimeoutError as cause:
            raise NomaTimeoutError(
                f"request exceeded {self._timeout_s}s", cause
            ) from cause


__all__ = [
    "DEFAULT_ARGS",
    "DEFAULT_COMMAND",
    "DEFAULT_TIMEOUT_S",
    "StdioMcpClient",
    "StdioMcpClientOptions",
    "ToolCallResult",
]
